features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/openbmp/BmpIntegrationAdapter.java

```diff
@@ -0,0 +1,724 @@
+                break;
+            case 3:
+                router.termReason = "Redundant connection.  The router has determined that this connection is redundant with another one";
+                break;
+            case 4:
+                router.termReason = "Session permanently administratively closed, will not be re-initiated";
+                break;
+            default:
+                router.termReason = "Unknown reason";
+        }
+
+        router.initData = null;
+        router.termData = termination.getMessage();
+        router.timestamp = context.timestamp;
+        router.bgpId = null;
+
+        this.handler.handle(new Message(context.collectorHashId, Type.ROUTER, ImmutableList.of(router)));
+    }
+
+    private void handlePeerUpNotification(final Transport.Message message,
+                                          final Transport.PeerUpPacket peerUp,
+                                          final Context context) {
+        final Transport.Peer bgpPeer = peerUp.getPeer();
+
+        final Peer peer = new Peer();
+        peer.action = Peer.Action.UP;
+        peer.sequence = sequence.getAndIncrement();
+        peer.name = InetAddressUtils.str(address(bgpPeer.getAddress())); // TODO: resolve Ip via DNS?
+        peer.hash = Record.hash(bgpPeer.getAddress(),
+                                bgpPeer.getDistinguisher(),
+                                context.routerHashId);
+        peer.routerHash = context.routerHashId;
+        peer.remoteBgpId = address(peerUp.getRecvMsg().getId());
+        peer.routerIp = context.sourceAddress;
+        peer.timestamp = timestamp(bgpPeer.getTimestamp());
+        peer.remoteAsn = uint32(peerUp.getRecvMsg().getAs());
+        peer.remoteIp = address(bgpPeer.getAddress());
+        peer.peerRd = asAttr((int) bgpPeer.getDistinguisher());
+        peer.remotePort = peerUp.getRemotePort();
+        peer.localAsn = uint32(peerUp.getSendMsg().getAs());
+        peer.localIp = address(peerUp.getLocalAddress());
+        peer.localPort = peerUp.getLocalPort();
+        peer.localBgpId = address(peerUp.getSendMsg().getId());
+        peer.infoData = peerUp.getMessage();
+        peer.advertisedCapabilities = ""; // TODO: Not parsed right now
+        peer.receivedCapabilities = ""; // TODO: Not parsed right now
+        peer.remoteHolddown = uint32(peerUp.getRecvMsg().getHoldTime());
+        peer.advertisedHolddown = uint32(peerUp.getSendMsg().getHoldTime());
+        peer.bmpReason = null;
+        peer.bgpErrorCode = null;
+        peer.bgpErrorSubcode = null;
+        peer.errorText = null;
+        peer.l3vpn = bgpPeer.getType() == Transport.Peer.Type.RD_INSTANCE;
+        peer.prePolicy = bgpPeer.getFlags().getPolicy() == Transport.Peer.Flags.Policy.PRE_POLICY;
+        peer.ipv4 = isV4(bgpPeer.getAddress());
+        peer.locRib = false; // TODO: Not implemented (see RFC draft-ietf-grow-bmp-loc-rib)
```

Comment:
TODO: Implement not or defer to JIRA?

Reply:
Deferred to https://issues.opennms.org/browse/NMS-12570